File: Examples/ObjectiveC/ObjectiveC/ViewController.m
//
//  ViewController.m
//  ObjectiveC
//
//  Created by Dave Duprey on 01/04/2021.
//

@import W3WSwiftComponents;
@import W3WSwiftApi;
#import <MapKit/MapKit.h>
#import "ViewController.h"


@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  // some coordinates for the component
  CGRect frame = CGRectMake(16.0, 64.0, self.view.frame.size.width - 32.0, 32.0);
  
  // make the autosuggest component
  W3WObjcAutoSuggestTextField *textfield = [[W3WObjcAutoSuggestTextField alloc] initWithFrame:frame];
  
  // assign the API key an set the desired language
  [textfield setApiKey:@"YourApiKey"];
  [textfield setLanguage:@"en"];
  
  // make focus and countryClipping options
  W3WObjcOptions *options = [[W3WObjcOptions alloc] init];
  [options addFocus: CLLocationCoordinate2DMake(51.520847,-0.195521)];
  [options addClipToCountry: @"GB"];
  [textfield setOptions: options];
  
  // add the component to the view
  [self.view addSubview:textfield];
  
  // print out any address that the user chooses
  [textfield setSuggestionCallback: ^(W3WObjcSuggestion *suggestion) {
    NSLog(@"%@", suggestion.words);
  }];
  
  // print out any error
  [textfield setErrorCallback: ^(NSError *error) {
    NSLog(@"Error: %@", error.localizedDescription);
  }];
  
}


@end

File: Tests/TestProjectTextField/TextField/ViewController.swift
///
//  ViewController.swift
//  TextField
//
//  Created by Dave Duprey on 26/11/2020.
//

import UIKit
import W3WSwiftApi
import W3WSwiftComponents
import CoreLocation
import Foundation


class ViewController: UIViewController {
    let testData = ClippingSettings()
  var apiKey: String = ""
      
    override func viewDidLoad() {
    super.viewDidLoad()

      if let apikey = ProcessInfo.processInfo.environment["PROD_API_KEY"] {
        self.apiKey = apikey
      } else if let apikey = getApikeyFromFile() {
        self.apiKey = apikey
      } else {
        print("Environment variable APIKEY must be set")
        abort()
      }

      
    let testData = ClippingSettings()

      var clipppingOptions : [W3WOption]
      {
          if let shape = UserDefaults.standard.string(forKey: "Clipping")
          {
              if let clipping = ClippingType(rawValue: shape)
              {
                  return testData.getClippingOptions(option : clipping)
              }
          }
          return testData.getClippingOptions(option : ClippingType.NoClipping)
      }
      
    let textField = W3WAutoSuggestTextField(frame: CGRect(x: 16.0, y: (UIApplication.shared.windows.first?.windowScene?.statusBarManager?.statusBarFrame.height ?? 64.0) + 16.0, width: view.frame.size.width - 32.0, height: 32.0))
    textField.accessibilityIdentifier = "w3wTextField"
               
    // assign the API to it
    let api = What3WordsV3(apiKey: apiKey)
    textField.set(api)
    textField.set(options: clipppingOptions)
    // turn on voice support
    textField.set(voice: true)
    textField.set(language: "en")

    // assign a code block to execute when the user has selected an address
    textField.onSuggestionSelected = { suggestion in
      print("User chose:", suggestion.words ?? "")
    }
      

    // the exact error can be captured using onError for whatever purpose you might have
    textField.onError = { error in
    self.showError(error: error)
    }
    
    // place in the view
    view.addSubview(textField)
  }

  func showError(error: Error) {
    DispatchQueue.main.async {
      let alert = UIAlertController(title: "Error", message: String(describing: error), preferredStyle: .alert)
      alert.addAction(UIAlertAction(title: "Dismiss", style: .cancel, handler: nil))
      self.present(alert, animated: true)
    }
  }
  
  
  func getApikeyFromFile() -> String? {
    var apikey: String? = nil
    
    let url = URL(fileURLWithPath: "/tmp/key.txt")
    if let key = try? String(contentsOf: url, encoding: .utf8) {
      apikey = key.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    return apikey
  }
  
}

